feat: link LTI Provider launches to authenticated users

tecoholic opened the pull request:
## Description

The LTI Provider feature of the platform currently allows accessing content from other LMS systems only as an anonymous user. Technically, there is support for using [LTI Authentication](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/lti/tpa_lti.html), which can be used for non-anonymous use case. However, this seems to have been built with [Blackboad integration in mind](https://openedx.org/blog/open-edx-lti-tool-provider/) _(see section "Authentication & User Provisioning")_ and doesn't work with Canvas when tested.

![image2](https://github.com/openedx/edx-platform/assets/383862/6dabd8c9-3fd7-4bcc-a445-ef23ce9aeaba)

Furthermore, the LTI Launch [view](https://github.com/openedx/edx-platform/blob/46bd8fb12fef49d67a0dc531416ed153f3414cfb/lms/djangoapps/lti_provider/views.py#L91C13-L91C13) and the [auth mechanism](https://github.com/openedx/edx-platform/blob/46bd8fb12fef49d67a0dc531416ed153f3414cfb/lms/djangoapps/lti_provider/users.py#L22) do not consider the existing logged-in user and [switch over to the anonymous user](https://github.com/openedx/edx-platform/blob/46bd8fb12fef49d67a0dc531416ed153f3414cfb/lms/djangoapps/lti_provider/users.py#L42) account when accessing content over LTI.

For organizations that use multiple LMS, this creates the issue of the grades and outcomes of users not being linked to their accounts despite having an account on the Open edX Platform.

This PR tries to bridge the gap by introducing an opt-in flag that can be set on the [LtiConsumer](https://github.com/openedx/edx-platform/blob/46bd8fb12fef49d67a0dc531416ed153f3414cfb/lms/djangoapps/lti_provider/models.py#L25) configurations that will allow linking existing users to the actions performed via LTI. The auto-linking of the `edx_user` and the `lti_user` is performed when the following conditions are met:
* The LTI Consumer performing the launch has the `Require user account` checkbox enabled.
* The LTI Consumer sends the `lis_person_contact_email_primary` during the LTI launch and an user account with the same email already exists in the platform.

**Who does this change affect?**
- Learners - The data generated via LTI will now be tied to their actual account on the platform instead of an anonymous account

**Screenshots**
A new flag is now added to the Admin UI of the LTI Consumer model.

<img width="724" alt="Screenshot 2023-11-06 at 4 54 04 PM" src="https://github.com/openedx/edx-platform/assets/383862/270c213d-bcfe-4b1c-91fd-07c311643688">


**Configuration Changes**

Since the LTI Provider is already behind the `ENABLE_LTI_PROVIDER` feature flag, and the auto-linking introduced here is only enabled with an explicit flag included in the PR, no other configuration has been introduced/changed.

## Supporting information

* [Discourse Discussion about the issue](https://discuss.openedx.org/t/open-edx-as-an-lti-provider-with-canvas/10014)
* [LTI 1.1 Specification](https://www.imsglobal.org/specs/ltiv1p1/implementation-guide) - Refer this for the `lis_person_contact_email_primary` optional parameter used in this PR to perform auto-linking.

## Testing instructions

**NOTE:** This has been tested only with Canvas LMS. Technically, it should be possible to test this with any LTI Consumer that sends the user's email using the `lis_person_contact_email_primary` parameter in the LTI launch request's POST data.

### Setting up the platform

1. Pull and switch to the PR branch in your devstack or Tutor dev setup.
2. [Enable the LTI Provider functionality](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/lti/enable_lti.html) 
3. [Create a new consumer](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/lti/configure_lti.html) and note down the client id and secret. **IMPORTANT:** Do NOT check the "Auto link users using email" check box.
4. Create 2 learner accounts
    * test_one@example.com - We will use this to test the existing functionality
    * test_two@example.com - We will use this to test the changes introduced in the PR
5. Now open any course on the LMS and grab the course ID and usage ID (attr `data-usage-id`) of an XBlock. 
<img width="825" alt="Screenshot 2023-09-26 at 12 08 03 PM" src="https://github.com/openedx/edx-platform/assets/383862/1c1468be-bfc4-4e65-b950-f019ed13eb69">


### Setting up Canvas 

* Follow the [Docker automated setup](https://github.com/instructure/canvas-lms/wiki/Quick-Start#setup) instructions and setup Canvas LMS _(requires cloning the repo, use --depth=1 while cloning to save some time and space)_
* Login to Canvas and create a new course
* Go to the Course Settings > Apps and click `+ App` to add a new app.
* In the form fill in the following details:
    - Configuration Type: _Manual Entry_
    - Name: _Any name you would like_
    - Consumer Key, Shared Secret: _LTI Consumer configuration values generated by the platform_
    - Launch URL: _Using the course and usage IDs noted before, enter the URL in the format:_ `http(s)://<your-domain>/lti_provider/courses/<course_id>/<usage_id>`
    - Privacy: _Email only_ or _Public_
* Add a new block to the course and choose **External Tool**. This should list the tool added in the previous step. It might take a couple of minutes or a forced refresh to get it listed as shown here. See Electric Circuit Tool.
<img width="684" alt="Screenshot 2023-09-26 at 12 14 44 PM" src="https://github.com/openedx/edx-platform/assets/383862/557af27f-4f07-4c5a-8d55-796d9dfa02d6">

* Publish the block and the course
* Add 2 users with the same email as the platform accounts as Students of the course from the **People** section of the settings. You will have to select the "User details" from the 3-dot menu -> Select "more details" -> "Add Login" under Login information -> set the username and password for the logins to work.

### Testing the changes

#### Without auto-linking (existing functionality)

* Go to the Open edX Platform's admin site -> Users and see the existing user list.
* In a private window, log into Canvas as the first test user `test_one` and accept the course assignment.
* Click the LTI Block in the course and verify that the embeded XBlock loads without any issue. _(A warning or error message may flash for a couple of seconds before the iframe is loaded)_
* Go to Open edX Admin Users list and refresh the page, there should be a new user with a random username like this: 
<img width="732" alt="Screenshot 2023-09-26 at 12 26 39 PM" src="https://github.com/openedx/edx-platform/assets/383862/0e97b4df-de70-4d8d-9b77-078473a2b045">

#### With auto-linking (introduced in this PR)

* In Open edX admin, go to **LTI Provider** > **Lti Consumers**, change the LTI Consumer created above and **check** the "Require user accounts" checkbox and save the configuration.
* Open another private window and log into Canvas using the second test user `test_two` and accept the course assignment, open the section with the XBlock.
* An message informing that the content is available only for signed in users will be displayed. <img width="802" alt="Screenshot 2023-11-07 at 4 15 24 PM" src="https://github.com/openedx/edx-platform/assets/383862/3638f717-5cbf-4993-890b-cb277ee9d2d1">
* Folow the instructions of the displayed message and verify that the XBlock loads as expected. Example:
    <img width="1280" alt="Screenshot 2023-09-26 at 12 34 25 PM" src="https://github.com/openedx/edx-platform/assets/383862/45c258aa-3a00-48e5-95e3-410713ef35aa">
* Now switch to the Open edX admin and go to the Users list. Note that no new users with random ID have been added as before.
* Open the LMS Shell and run `./manage.py lms shell`
* Execute the following code block to verify that the second LTI User has been actually linked to the existing user account.

```py
>> from lms.djangoapps.lti_provider.models import LtiUser
>> lti_users = LtiUser.objects.all()
>> print([lti_user.edx_user.email for lti_user in lti_users])
['OnlzRDMypJLBOGIytSSDImjZB03oTf@lti.example.com', 'test2@example.com']
```

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

* At the outset, this kind of auto-linking using an OPTIONAL field `lis_person_contact_primary_email` feels like unnecessary as addressing the issues with the broken LTI Authentication mechanism mentioned earlier in the description should have this working out of the box. However, it was discovered during the investigation that the `LtiUser` in only ever created from the `lti_launch` view (also noted by @ztraboo in the [forum comment here](https://discuss.openedx.org/t/open-edx-as-an-lti-provider-with-canvas/10014/5)). This means, in order for the auto-linking to work, on top of fixing the LTI Authentication, we will still need to modify the `LtiUser` creation logic, which this PR handles.
* [Related PR for updating the documentation](https://github.com/openedx/edx-documentation/pull/2190)

tecoholic committed:
feat: adds auto_link_users_using_email field to LtiConsumer

openedx-webhooks commented:
Thanks for the pull request, @tecoholic! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

tecoholic committed:
feat: link existing users to the LtiUser via email

With this change the platform users who access content via LTI will
automatically be linked their platform account instead of the anonymous
account when the following conditions are met:
* the LtiConsumer should be configured to auto link the users via email
* the LTI Consumer should share the email of the user using the
lis_person_contact_email_primary parameter in the LTI Launch POST data

Internal-ref: https://tasks.opencraft.com/browse/BB-7875

Cup0fCoffee commented on lms/djangoapps/lti_provider/users.py:
```diff
@@ -44,34 +47,39 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
+            except IntegrityError:
+                # The random edx_user_id wasn't unique. Since 'created' is still
+                # False, we will retry with a different random ID.
+                pass
```
@tecoholic
```suggestion
    def get_or_create_edx_user(email):
        if email:
            edx_user = User.objects.filter(email=email).first()

        if not edx_user:
            created = False
            edx_password = str(uuid.uuid4())
            while not created:
                try:
                    edx_user_id = generate_random_edx_username()
                    edx_email = f"{edx_user_id}@{settings.LTI_USER_EMAIL_DOMAIN}"
                    with transaction.atomic():
                        edx_user = User.objects.create_user(
                            username=edx_user_id,
                            password=edx_password,
                            email=edx_email,
                        )
                        # A profile is required if PREVENT_CONCURRENT_LOGINS flag is set.
                        # TODO: We could populate user information from the LTI launch here,
                        # but it's not necessary for our current uses.
                        edx_user_profile = UserProfile(user=edx_user)
                        edx_user_profile.save()
                        created = True
                except IntegrityError:
                    # The random edx_user_id wasn't unique. Since 'created' is still
                    # False, we will retry with a different random ID.
                    pass

         return edx_user

     edx_user = get_or_create_edx_user(email)
```

Note: I've copied the `edx_password = str(uuid.uuid4())` from outside the selected scope, so you wound need to remove it from above, if you decide to accept this change.

tecoholic replied:
@Cup0fCoffee Because of the way you wrote this function, now a new use case jumped to my mind. What happens when we encounter an email in the LTI request, but the user doesn't have an account? The current implementation would create an anonymous user with a random ID. And even after they explicitly create an account on the platform, they would still end up using the anonymous account because the `authenticate_lti_user` function only queries the user with `lti_user_id` and `lti_consumer`. Which means, the auto-linking will only be possible for the specific users, who have an account on the Open edX platform and are using LTI for the first time using that specific LTI Consumer credentials. 

I am not sure what a good default here is, creating an account on behalf of someone else on the platform is probably a PII issue. So we can't create a new user with the `lis..email` every time we get an email. Which also means we this flag should be set only during the LtiConsumer creation, changing it later will not do the intended job of linking learners who are already associated with an anonymous account.

Ah! This is suddenly more complicated than before. 😛

tecoholic replied:
After some thinking
* creating accounts for users with email is definitely a bad idea. Depending on the auth setup like SSO, this can have unintended consequences.
* allowing the flag to be changes back and forth will create the issue of having some users associated with their actual account and others with an anonymous account. This will cause confusion. So, the flag should only set when the `LtiConsumer` is created and can't be unset. If, for some reason, the auto-linking is no longer desired, then a new 
`LtiConsumer` should be created with the flag off.
* Finally, we will document that this auto-linking only works when the learners already have an account on the platform and can't be used to retrospectively link learners.

tecoholic replied:
@Cup0fCoffee I implemented your suggestion but, without defining a new function. I hope that's okay :)

Cup0fCoffee commented on lms/djangoapps/lti_provider/users.py:
```diff
@@ -35,7 +35,10 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
+        lis_email = ""
+        if lti_consumer.auto_link_users_using_email:
+            lis_email = request.POST.get("lis_person_contact_email_primary", "")
+        lti_user = create_lti_user(lti_user_id, lti_consumer, lis_email)
```
@tecoholic I don't like that we always pass `lis_email`, because if we were to change the default value in the function signature, let say to `email=None`, in this call the default value will always be overwritten to the empty string. I understand the argument of this being concise, but it can have negative impact long term. Also, imho, semantically `None` is better as default value than empty string in this case.

tecoholic replied:
@Cup0fCoffee This is an interesting one. I have started using type hints in the new code I write that I have gotten into the habit of using empty strings instead of `None` to avoid the `str|None` in the function definitions. And as a boolean, it works out the same. That has made me use empty strings in more places than I probably should.

However, I think you are right, here the not sending that argument is a better way to communicate the absence of an email instead of an empty string.

Cup0fCoffee commented:
> Launch URL: Using the course and usage IDs noted before, enter the URL in the format: `http(s)://<your-domain>/lti_provider/courses/<course_id>/<usage_id>/`

Didn't work for me with the trailing slash. After I removed, it worked.

Cup0fCoffee commented:
> Open the LMS Shell and run `./manage.py shell`

`./manage.py lms shell`

Cup0fCoffee reviewed (approved):
:+1: Once the other comments are addressed. Also, is there someone this can be documented for admins?

- [x] I tested this: followed the testing instructions.
- [x] I read through the code
- [-] I checked for accessibility issues
- [ ] Includes documentation

tecoholic committed:
refactor: use None as a default value in create_lti_user

tecoholic committed:
feat: make the auto link flag editable only during object creation

tecoholic committed:
fix: remove the empty strings as args from the tests

tecoholic commented:
@Cup0fCoffee I have added the documentation to the [relevant section](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/lti/configure_lti.html) and create a [PR](https://github.com/openedx/edx-documentation/pull/2190).

ztraboo commented on lms/djangoapps/lti_provider/users.py:
```diff
@@ -44,34 +48,36 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
+                pass
 
     lti_user = LtiUser(
         lti_consumer=lti_consumer,
```
@tecoholic 
I noticed that the `UNIQUE KEY `edx_user_id` (`edx_user_id`),` will fail if you have multiple LTI Consumer connections from the same email on two separate LTI Consumer instances.

Probably need to remove this unique key constraint on the database to fix this issue. This is because the `edx_user` id is found based on email.

There may not be a case for this because most likely a user won't be enrolled at two separate LMS institutions with the same email address. Let me know what you think about this.
```
CREATE TABLE `lti_provider_ltiuser` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `lti_user_id` varchar(255) NOT NULL,
  `edx_user_id` int(11) NOT NULL,
  `lti_consumer_id` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `edx_user_id` (`edx_user_id`),
  UNIQUE KEY `lti_provider_ltiuser_lti_consumer_id_lti_user_id_d684c543_uniq` (`lti_consumer_id`,`lti_user_id`),
  CONSTRAINT `lti_provider_ltiuser_edx_user_id_e06b2db5_fk_auth_user_id` FOREIGN KEY (`edx_user_id`) REFERENCES `auth_user` (`id`),
  CONSTRAINT `lti_provider_ltiuser_lti_consumer_id_c1b42ee3_fk_lti_provi` FOREIGN KEY (`lti_consumer_id`) REFERENCES `lti_provider_lticonsumer` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2011 DEFAULT CHARSET=latin1Ω
```

tecoholic replied:
@ztraboo Thanks for catching this edge case. This will be a real issue where multiple consumers are configured – for example, Canvas allows LTI Tools to be added per course (so does Open edX actually). So, there could be multiple courses configured as independent LTI Consumers in the same instance. This will lead to `LtiUser` creation failing due to this `UNIQUE` constraint. So maintaining the combination of uniqueness of `lti_consumer_id` and the `lti_user_id` should be sufficient. Expecting an `edx_user` to use only one LTI User will not work with the auto-linking. I will update the PR to handle this.

tecoholic replied:
@ztraboo Replaced the one-to-one relationship with a one-to-many in [this commit](https://github.com/openedx/edx-platform/pull/33310/commits/ca81ca562dc86ade17e45551c1a5cec8b1fb0589). Now multiple `LtiUser`s can be linked to the same `edx_user`.

tecoholic committed:
fix: remove unique constraint on edx_user

With the auto linking of edx_user with the lti_users, the scenario where
multiple LTI consumers will create independent LtiUsers depending on the
same edx_user is created. This will not be possible if the edx_user has
a one-to-one relationship with the lti_user. This commit replaces the
one-to-one relationship with an one-to-many relationship so that
multiple LtiUser objects can be created referencing the same edx_user.

Agrendalath reviewed (commented):
@tecoholic, it took me a while to set up Canvas and learn the way around it. Two main issues I encountered while testing this:
1. Permission errors after starting Canvas. ([solution](https://github.com/instructure/canvas-lms/tree/4fbc04c/doc/docker))
2. Canvas generates identical GUIDs for the Open edX LTI Consumers. The GUID has precedence over the client_id while retrieving the correct consumer, so I had to manually remove this value from the consumer after each LTI launch.

I like this solution - it's clean and straightforward. **But...**

_Assumption: none of your Canvas users has staff or superuser permissions in Open edX._
Let's say somebody gains unauthorized access to your admin account of the Canvas instance. They can modify the email address to be **anything** without any confirmation. This way, they can send any value in the `lis_person_contact_email_primary`, thus impersonating any user in the Open edX.

This _might_ be an acceptable risk for instances managed by a single organization, but it can be dangerous for multi-org instances.

For a comparison, when you try to use SAML or OAuth2 with an email already bound to an Open edX account, you are redirected to the login page. When you log in (with your password or an already confirmed TPA), the `UserSocialAuth` account is linked to your main account.
Could we utilize the `LTIProviderConfig` from the `third_party_auth` app to use this existing mechanism?

tecoholic commented:
@cmltaWt0 Hi, we were trying to get this into Quince. But discovered a potential security issue and are working on updating the PR to handle it. I will create a backport and add it to the [backport list](https://github.com/openedx/wg-build-test-release/issues/307) once this goes through. Just wanted to give you a heads-up and see if you have any concerns.

tecoholic committed:
refactor: remove read-only from Lti Consumer config admin

tecoholic committed:
fix: allow only authenticated users to be auto-linked

tecoholic committed:
fix: extra space linting issue

tecoholic commented:
@Agrendalath Hi, after exploring multiple avenues, I think I have a fair solution here. The important changes are:

1. The `LtiConsumer` object is editable – the auto-linking switch can be toggled. This should get rid of the issue of creating duplicate `LtiConsumer`s from the same instance, which were causing the instance GUID clashes.
2. The auto-linking is done only when `request.user.email` and the `lis_person_contact_email_primary` match. If the user is not-authenticated or authenticated with a user account with different email, the session will be cleared, and an error message is shown. <img width="1191" alt="User Auth Error in LTI Consumer" src="https://github.com/openedx/edx-platform/assets/383862/31fd7cde-66b3-4bee-b9d2-124c6de340f4">
3. If an LtiUser object already exists with an anonymous edX User, it will be switched over to the “real account” with the matching email. This is done to maintain the lti_user_id + lti_consumer unique constraint. One side effect of this is, once the LtiUser and the edX user are linked, going back to an anonymous user is not possible on the same LtiConsumer, a new LtiConsumer will have to be created.
4. Finally, the auto-linking will work only if the `SESSION_COOKIE_SAMESITE` value is set to `'None'` and `SESSION_COOKIE_SECURE` is set to `True` as the iframes on the LTI Consumer will need it to share the session. If it is not desirable, then the LTI Launch should happen in a new tab instead of the IFrame.

### Testing Tips:
1. Use Chrome or Chromium - [localhost sites behave like HTTPS](https://web.dev/articles/how-to-use-local-https) and the `SameSite=None;Secure` cookies are forwarded as expected in a prod-env with HTTPS. I lost a good full day trying to make Firefox do the same, but didn't succeed.
2. Use devstack instead of Tutor, as tutor serves from `local.overhang.io` and making it run in HTTPS on dev setup can be challenging. Using a http proxy in front of it, didn't work as the OAuth signature validation kept failing. Maybe there is a way to overcome both here, but devstack provides a more straightforward solution.
3. My `lms/envs/private.py` used for testing:
```py
from .common import FEATURES, INSTALLED_APPS, AUTHENTICATION_BACKENDS

FEATURES['ENABLE_LTI_PROVIDER'] = True
INSTALLED_APPS.append('lms.djangoapps.lti_provider.apps.LtiProviderConfig')
AUTHENTICATION_BACKENDS.append('lms.djangoapps.lti_provider.users.LtiBackend')

SESSION_COOKIE_SAMESITE = 'None'
SESSION_COOKIE_SECURE = True
SESSION_COOKIE_SAMESITE_FORCE_ALL = True
CSRF_COOKIE_SECURE = True
CSRF_COOKIE_SAMESITE = 'None'
# Needed for showing pages in iframe
X_FRAME_OPTIONS = "ALLOW-FROM canvas.docker"

# These are probably not needed
ENABLE_CORS_HEADERS = True
ENABLE_CROSS_DOMAIN_CSRF_COOKIE = True

DCS_SESSION_COOKIE_SAMESITE = 'None'
DCS_SESSION_COOKIE_SECURE = True
DCS_SESSION_COOKIE_SAMESITE_FORCE_ALL = True
```

Kindly review and see if the security concerns are addressed. I might have inadvertently created more with the cookie permissions, but they can be mitigated by launching on new tab. But everything have been tested in iframe to ensure we are not forcing a new tab on the learner for anything apart from the "Sign in".

kaustavb12 commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33310-139931.staging.do.opencraft.hosting](https://pr-33310-139931.staging.do.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33310-139931.staging.do.opencraft.hosting](https://studio.pr-33310-139931.staging.do.opencraft.hosting)

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at [pr-33310-139931.staging.do.opencraft.hosting](https://pr-33310-139931.staging.do.opencraft.hosting)
Sandbox Studio is available at [studio.pr-33310-139931.staging.do.opencraft.hosting](https://studio.pr-33310-139931.staging.do.opencraft.hosting)

Agrendalath commented on lms/djangoapps/lti_provider/users.py:
```diff
@@ -35,7 +35,21 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
+        request.user.is_authenticated and
+        lti_user.edx_user != request.user
+    ):
+        print("Switching to request.user")
```
:)

Agrendalath commented on lms/djangoapps/lti_provider/views.py:
```diff
@@ -86,6 +87,18 @@ def lti_launch(request, course_id, usage_id):
     params['course_key'] = course_key
     params['usage_key'] = usage_key
 
+    # Verify that the email from the LTI Launch and the logged-in user are the same.
```
I think this logic shouldn't be handled by a view but directly in the `authenticate_lti_user`. We can raise an exception when this happens and then catch it + render the error in the view.

tecoholic replied:
@Agrendalath Update the PR with the suggested change.

Agrendalath reviewed (commented):
@tecoholic, I'm trying to understand why we need to change these cookie settings now. [The previous approach](https://github.com/openedx/edx-platform/pull/33310/commits/eb6f6bf7b4a4b863debe09c6c1f564cd3dadaf6e) worked correctly without these changes. What did we change that requires this? Does it mean that the [user switching functionality](https://github.com/openedx/edx-platform/pull/33310/commits/eb6f6bf7b4a4b863debe09c6c1f564cd3dadaf6e#diff-ca066e926cd72b473ba83ba6217e0d21d9f24fda0da8ceccdddadf7c2cdebca5R43-R47) has not been working correctly since the default cookie behavior was altered in the browsers?

> didn't work as the OAuth signature validation kept failing.

Maybe it's not the best approach, but I usually just turn this validation off by making the validation functions return `True` in the devstack.

tecoholic committed:
refactor: move the email comparison check to auth function

tecoholic committed:
fix: apply code deletion missed in last commit

open-craft-grove commented:
Sandbox update request received. Deployment will start soon.

tecoholic committed:
fix: flush the session before showing the error message

tecoholic commented:
@Agrendalath 
> @tecoholic, I'm trying to understand why we need to change these cookie settings now. [The previous approach](https://github.com/openedx/edx-platform/pull/33310/commits/eb6f6bf7b4a4b863debe09c6c1f564cd3dadaf6e) worked correctly without these changes. What did we change that requires this? Does it mean that the [user switching functionality](https://github.com/openedx/edx-platform/pull/33310/commits/eb6f6bf7b4a4b863debe09c6c1f564cd3dadaf6e#diff-ca066e926cd72b473ba83ba6217e0d21d9f24fda0da8ceccdddadf7c2cdebca5R43-R47) has not been working correctly since the default cookie behavior was altered in the browsers?

The user-switching was working before because, the `LtiUser` was fetched with just the `user_id` from the LTI Launch param and then the anonymous `edx_user` was authenticated using the [`LtiBackend`](https://github.com/open-craft/edx-platform/blob/253a45d89da35eff6818eaffe6e86e57ee3331b0/lms/djangoapps/lti_provider/users.py#L139-L140) - which only required that the `lti_user_id` had an object with a connected `edx_user`. So even when the session cookie wasn't being sent due to the new browser implementations in the requests, the responses had an established session inside the Iframe. So, in a way the `switch_user` was both working and not working.
* It was working by switching the `AnonymousUser` with an actual `edx_user`
* It wasn't working as in, it didn't switch 2 different `edx_users`

We need to change the cookie settings now, as we need to be able to authenticate the user who logged in outside the Iframe.

tecoholic committed:
fix: linting issue

open-craft-grove commented:
Sandbox deployment failed.

Check failure logs here [https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/34602668-5445942825.log](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/34602668-5445942825.log)

Please check the settings and requirements and retry deployment by updating the pull request or posting a `/grove sandbox update` comment.

open-craft-grove commented:
Sandbox deployment started.

open-craft-grove commented:
Sandbox deployment failed.

Check failure logs here [https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/34602668-5446021602.log](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/34602668-5446021602.log)

Please check the settings and requirements and retry deployment by updating the pull request or posting a `/grove sandbox update` comment.

open-craft-grove commented:
Sandbox deployment failed.

Check failure logs here [https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/34602668-5446113570.log](https://grove-stage-build-logs.nyc3.digitaloceanspaces.com/34602668-5446113570.log)

Please check the settings and requirements and retry deployment by updating the pull request or posting a `/grove sandbox update` comment.

Agrendalath commented on lms/djangoapps/lti_provider/users.py:
```diff
@@ -28,14 +28,35 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
+
+    # Verify that the email from the LTI Launch and the logged-in user are the same.
+    if lti_consumer.auto_link_users_using_email and (
+        not request.user.is_authenticated or
```
> The auto-linking is done only when request.user.email and the lis_person_contact_email_primary match. **If the user is not-authenticated** or authenticated with a user account with different email, the session will be cleared, and an error message is shown. 

This is different from the initial approach. Do we want to force users to create an account when using this option? If yes, perhaps we should create a choice field with "the default approach", "auto-linking", and "auto-linking with registration requirement" (we should think about better names for these). Alternatively, we could change the name of this field and add some helper text to indicate the implications.

tecoholic replied:
Yes, we would like to make sure all the users have an account on Open edX before they access the content via LTI. This is an intentional change to maintain consistency. With the earlier implementation, users with and without an account had different experiences. Now everyone has the same experience and they are all expected to have an account.

To reflect this, I have renamed the flag to `require_user_account` and added help text explaining the outcome of enabling this flag. The [documentation has been updated](https://github.com/openedx/edx-documentation/pull/2190/files) with the relevant details as well.

Agrendalath replied:
@tecoholic, got it. However, the current implementation does not work in the following scenario:
1. Log into Open edX.
2. Use LTI to link your account (i.e., create the `LtiUser`).
3. Log out from Open edX.
4. Use LTI again. Currently, you need to log into Open edX again. However, the `LtiUser` is already created (the accounts are linked), so learners should not need to sign into Open edX anymore.

tecoholic replied:
@Agrendalath Hmm. I thought requiring that the Open edX user being logged in made it avoid the security issue of spoofed emails. So I didn't see the need to handle the user being logged out. 

I am also wondering how to implement the email validation when we don't have an `request.user`.  Perhaps use the `lti_user_id` which is supposed to be unique and then throw an error if the `lti_user.edx_user.email` doesn't match the `lis_email`? 🤔 But that would also mean, sometimes the `edx_user` could be the anonymous account (flag turned on after a learner has used it anonymously) and their email would never match. Then we will have to filter for the `lti.example.com` in the email to identify anonymous vs real accounts. I wonder if this level checking is worth it.

Is requiring the users being logged in to both the systems a bad requirement?

Agrendalath replied:
@tecoholic,
> Is requiring the users being logged in to both the systems a bad requirement?

I believe so. One of the major benefits of using Open edX as an LTI provider is the seamless authentication from another LMS. Otherwise, we could add an annotation that you need to be logged in to access the content (provide a link to the Open edX login page), and just embed the `/xblock/{block_id}` view in an iframe below.

> Hmm. I thought requiring that the Open edX user being logged in made it avoid the security issue of spoofed emails.

Not really - we care about this issue only when linking an account. My initial explanation about the vulnerability might have been a bit fuzzy. Let's compare this to the OAuth2 mechanism, which is very similar in this aspect.

It consists of the SP (our instance) and the IdP (external service). Let's say we use site A as the IdP (_I wanted to call it X, but it could be controversial, given the recent branding change of one of the bigger sites_). If we fully trust A before linking the user account, they can provide any email to take over any account (e.g., one with the superuser permissions) on our instance. Therefore, we don't want to fully trust them with linking the users. Instead, it's the user who should decide if they want to trust this site to act as an IdP on their behalf. When they link the account, they can log into the platform without a password when they are authenticated on A. However, if A decides to provide a fake email address, they will not be able to access the superuser who does not have the account linked with their service. Therefore, _we_ are not trusting them with our whole user base - instead, it's the _individual users_ who trust them with their accounts.

Back to the LTI - since the `LtiUser` is tied to a specific `LtiConsumer`, there is no way for the LTI consumer to take over an account of a user who did not link the account with this service first. This way, we limit the attack surface to users who decide to use their system, without sacrificing the major benefit of using another LMS as the seamless IdP.

> Perhaps use the lti_user_id which is supposed to be unique (...)

Yes - [it is unique](https://github.com/openedx/edx-platform/blob/390c2fb8ca0ef068d199587cf302551f8d2459ab/lms/djangoapps/lti_provider/models.py#L151-L152) for the `lti_consumer` + `lti_user_id` pair.

> I am also wondering how to implement the email validation when we don't have an request.user.

We don't need to check the email. We already get the `LtiUser` [here](https://github.com/openedx/edx-platform/blob/390c2fb8ca0ef068d199587cf302551f8d2459ab/lms/djangoapps/lti_provider/users.py#L41-L44), based on the `lti_user_id` ([retrieved from the LTI launch params](https://github.com/openedx/edx-platform/blob/5b464b6a0b078c0a55c6fba33276c22d74fde588/lms/djangoapps/lti_provider/views.py#L94)) and `lti_consumer` ([retrieved from our DB, with the GUID/key validation](https://github.com/openedx/edx-platform/blob/5b464b6a0b078c0a55c6fba33276c22d74fde588/lms/djangoapps/lti_provider/views.py#L66-L69)).

> and then throw an error if the lti_user.edx_user.email doesn't match the lis_email? 🤔

We don't care about it at this point. The `LtiUser` already exists, which means that the user has already linked their account.

> But that would also mean, sometimes the edx_user could be the anonymous account (flag turned on after a learner has used it anonymously) and their email would never match. Then we will have to filter for the lti.example.com in the email to identify anonymous vs real accounts. I wonder if this level checking is worth it.

As we don't care about the email for the already linked accounts, we do not need to handle this scenario.

tecoholic replied:
@Agrendalath Thank you for explaining everything in detail. The explanation on balancing trust between the entities was really helpful. 

I will work on the changes to handle the missing scenario and ping you again for review.

tecoholic replied:
@Agrendalath I have fixed the issue and tested that LTI launch will succeed after the user is logged out. I also added a test case to verify this in 78c5b9c47bcf2a2f1e9eec54e5404fc69c521d86.

Agrendalath commented on lms/templates/lti_provider/user-auth-error.html:
```diff
@@ -0,0 +1,24 @@
+    ${Text(_("This module is available only for users who are signed into {platform}. "
+        "Kindly sign in and refresh this page to load the content."
+      )).format(
+        platform=HTML(u"<em>{}</em>").format(static.get_platform_name())
```
```suggestion
        platform=HTML("<em>{}</em>").format(static.get_platform_name())
```

Nit: fortunately, we already use Unicode :)

Agrendalath reviewed (commented):
@tecoholic, all good from my end. I just want to make sure that we didn't mistakenly change the requirement to enforce user registration.

<br/>

> Use Chrome or Chromium - [localhost sites behave like HTTPS](https://web.dev/articles/how-to-use-local-https) and the SameSite=None;Secure cookies are forwarded as expected in a prod-env with HTTPS. I lost a good full day trying to make Firefox do the same, but didn't succeed.

For testing, you can do something like:
```python
class IframesForEveryoneMiddleware(MiddlewareMixin):
    def process_response(self, request, response):
        response['X-Frame-Options'] = 'ALLOWALL'
        return response
```

Then, just add it to the middlewares somewhere after the `'django.middleware.clickjacking.XFrameOptionsMiddleware',`.


<br/><br/>

> We need to change the cookie settings now, as we need to be able to authenticate the user who logged in outside the Iframe.

Agreed. I thought about doing something like this to verify the user instead:
```python
from edx_rest_framework_extensions.auth.jwt.cookies import (
    jwt_cookie_header_payload_name,
    jwt_cookie_name,
    jwt_cookie_signature_name,
)
from edx_rest_framework_extensions.auth.jwt.decoder import configured_jwt_decode_handler

header = request.COOKIES.get(jwt_cookie_header_payload_name())
signature = request.COOKIES.get(jwt_cookie_signature_name())
jwt_cookie = request.COOKIES.get(jwt_cookie_name()) or f"{header}{JWT_DELIMITER}{signature}"
if jwt_cookie:
    payload = configured_jwt_decode_handler(jwt_cookie)
    if payload.get("email_verified") and payload.get("email") == lis_email:
        return True
```

It worked well in Firefox (it has the "SameSite" of JWT cookies set to "None"), but then I noticed that Chromium does have a value for the "SameSite" there (so it probably uses "Lax").

tecoholic committed:
refactor: use a better name for the flag with helptext

tecoholic commented:
@Agrendalath Thank you for the detailed review. I have updated this PR and the related [documentaiton PR](https://github.com/openedx/edx-documentation/pull/2190/files) based on your feeback.

> Agreed. I thought about doing something like this to verify the user instead:

I wanted to use the JWT Cookies as well. It somehow seemed more secure than forwarding session cookies through iframe. However, I noticed that the signature cookie had different flags set than the payload cookie. So I didn't go down that route as I thought it could be more complex. Also I had poorer understanding of this whole cookie situation than now. Thanks for the snippet. Now I have a reference of how to use them in the future.

Agrendalath commented on lms/djangoapps/lti_provider/migrations/0004_require_user_account.py:
```diff
@@ -0,0 +1,26 @@
+        migrations.AddField(
+            model_name='lticonsumer',
+            name='require_user_account',
+            field=models.BooleanField(blank=True, default=False, help_text='When checked, the LTI content will load only for learners who have an account in this instance. This is required only for linking leaner accounts with the LTI consumer. See the Open edX LTI Provider documentation for more details.'),
```
```suggestion
            field=models.BooleanField(blank=True, default=False, help_text='When checked, the LTI content will load only for learners who have an account in this instance. This is required only for linking learner accounts with the LTI consumer. See the Open edX LTI Provider documentation for more details.'),
```

Without this, Django complains about changes that are not reflected in a migration.

Agrendalath commented on lms/templates/lti_provider/user-auth-error.html:
```diff
@@ -0,0 +1,24 @@
+    ${Text(_("There was an error when loading this module!"))}
+  </h1>
+  <p>
+    ${Text(_("This module is available only for users who are signed into {platform}. "
+        "Kindly sign in and refresh this page to load the content."
```
We should mention that the account needs to be associated with the same email on both platforms, right? Or do we want to allow linking the LTI user to the one who is currently logged into Open edX?

This second option might be more useful and intuitive.

tecoholic replied:
@Agrendalath Right now the whole linking depends on the email being sent in the LTI launch request (`lis_person_contact_email_primary`). So keeping it restricted to allow this only when emails match might be the best. While not intuitive, we can also avoid scenarios like shared computers in college libraries where the new LTI user gets connected to the previous edX user.

I will update the message to explicitly mention that emails should match.

tecoholic replied:
Updated the error page in 0642dd7a8bab93500757de467bd3ea4bb6c6ae9f

tecoholic committed:
fix: typo in migration help text

Co-authored-by: Piotr Surowiec <piotr@surowiec.it>

tecoholic committed:
fix: explicitly mention emails should match in the error page

tecoholic committed:
fix: lti launch failing when auto-linked user was logged out of platform

tecoholic committed:
fix: linting issue with raising exception

Agrendalath commented on lms/djangoapps/lti_provider/users.py:
```diff
@@ -28,14 +28,34 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
+        lti_user.edx_user != request.user
+    ):
+        lti_user.edx_user = request.user
+        lti_user.save()
```
@tecoholic, let's assume we have Alice and Bob who are the learners. Is the following scenario correct?
1. Alice has already used LTI, so her Canvas and Open edX accounts are linked.
2. Bob used a shared computer in a classroom and left his Open edX account logged in.
3. Alice uses the same shared computer, logs into Canvas, and goes to a Canvas page with the LTI content from Open edX.
4. We have one of the following outcomes:
    1.  If Bob had used LTI before, this raises an `IntegrityError` (due to the `'lti_consumer', 'lti_user_id'` uniqueness violation).
    2. Otherwise, Bob's account is accidentally "hijacked" by Alice.
        1. Bob's account will get re-linked again (to the correct account) the next time he uses the LTI, but it is still a security concern.

In this scenario, we want to switch the Open edX user to the one linked to the `LtiUser` (with the `switch_user` function).
The only real-world usage I can think of for re-linking the accounts is moving the existing (automatically created) LTI users to "real" Open edX accounts after enabling an existing provider's `require_user_account`  option. If this is our goal, we should verify if the `User`'s email ends with `settings.LTI_USER_EMAIL_DOMAIN`. However, all existing progress of such users will be lost this way, so we should be sure that we want to do this (instead of, e.g., changing the emails of their automatically created accounts).

tecoholic replied:
@Agrendalath First up, I am just astounded at how many of the issues I miss and you catch. And I entirely understand that's the function of the work-review cycle and really appreciate it.

And yes, the only real reason to switch the user here is to make sure we don't have a mix of real users and anonymous users on a client if/when the flag is turned on. I had my reservations about it, but this is something I totally missed

> However, all existing progress of such users will be lost this way, so we should be sure that we want to do this 

So, I think it is best to remove the re-linking part now. I will update the PR.

tecoholic committed:
fix: remove re-linking from authenticate_user function

Agrendalath reviewed (approved):
Great work, @tecoholic!

:+1:

* [x] I tested this: linking existing Open edX accounts with the ones from LTI
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation: https://github.com/openedx/edx-documentation/pull/2190

openedx-webhooks commented:
@tecoholic 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/33310&entry.1484655318=https://github.com/tecoholic&entry.752974735=2023-09-21+12:28&entry.1917517419=2023-11-13+18:38&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
